spec/abstract/AARCH64/ArchVSpace_A.thy

```diff
@@ -0,0 +1,686 @@
+definition get_vmid :: "asid \<Rightarrow> (vmid, 'z::state_ext) s_monad" where
+  "get_vmid asid \<equiv> do
+     maybe_vmid \<leftarrow> load_vmid asid;
+     case maybe_vmid of
+       Some vmid \<Rightarrow> return vmid
+     | None \<Rightarrow>  do
+         new_hw_asid \<leftarrow> find_free_vmid;
+         store_vmid asid new_hw_asid;
+         return new_hw_asid
+       od
+   od"
+
+text \<open>
+  Format a VM fault message to be passed to a thread's supervisor after it encounters a page fault.\<close>
+definition handle_vm_fault :: "obj_ref \<Rightarrow> vmfault_type \<Rightarrow> (unit, 'z::state_ext) f_monad" where
+  "handle_vm_fault thread fault \<equiv> case fault of
+     ARMDataAbort \<Rightarrow> doE
+       addr \<leftarrow> liftE $ do_machine_op getFAR;
+       fault \<leftarrow> liftE $ do_machine_op getDFSR;
+       cur_v \<leftarrow> liftE $ gets (arm_current_vcpu \<circ> arch_state);
+       addr \<leftarrow> if (\<exists>v. cur_v = Some (v, True)) \<comment> \<open>VCPU active\<close>
+              then doE
+                  \<comment> \<open>address bits of PAR register after S1 translation\<close>
+                  par_el1_mask \<leftarrow> returnOk $ 0xfffffffff000;
+                  addr' \<leftarrow> liftE $ do_machine_op $ addressTranslateS1 addr;
+                  returnOk $ (addr' && par_el1_mask) || (addr && mask pageBits)
+                odE
+              else returnOk addr;
+       \<comment> \<open>32 is the width of the FSR field in the C VMFault structure\<close>
+       throwError $ ArchFault $ VMFault addr [0, fault && mask 32]
+     odE
+   | ARMPrefetchAbort \<Rightarrow> doE
+       pc \<leftarrow> liftE $ as_user thread $ getRestartPC;
+       fault \<leftarrow> liftE $ do_machine_op getIFSR;
+       cur_v \<leftarrow> liftE $ gets (arm_current_vcpu \<circ> arch_state);
+       pc \<leftarrow> if (\<exists>v. cur_v = Some (v, True)) \<comment> \<open>VCPU active\<close>
+            then doE
+                \<comment> \<open>address bits of PAR register after S1 translation\<close>
+                par_el1_mask \<leftarrow> returnOk $ 0xfffffffff000;
+                pc' \<leftarrow> liftE $ do_machine_op $ addressTranslateS1 pc;
+                returnOk $ (pc' && par_el1_mask) || (pc && mask pageBits)
+              odE
+            else returnOk pc;
+       \<comment> \<open>32 is the width of the FSR field in the C VMFault structure\<close>
+       throwError $ ArchFault $ VMFault pc [1, fault && mask 32]
+     odE"
+
+
+text \<open>Switch to given address space, using VMID associated with provided ASID.\<close>
+definition arm_context_switch :: "obj_ref \<Rightarrow> asid \<Rightarrow> (unit,'z::state_ext) s_monad" where
+  "arm_context_switch vspace asid = do
+     vmid <- get_vmid asid;
+     do_machine_op $ setVSpaceRoot (addrFromPPtr vspace) (ucast vmid)
+   od"
+
+
+section \<open>Manipulation of VCPU-related state and registers\<close>
```

Comment:
That might be nice, I'll have a look.

Reply:
That seems to be working nicely. There is a bit of a question of what should go into `VCPUAcc_A` and what into `VCPU_A`. For now I've erred on the side of keeping them on roughly the same heigh in the theory graph in `VCPUAcc_A`, but we can move more whenever we feel like it.

Reply:
Cool, good to see that it could be factored out so nicely!
Very low priority, but we should do the same for ARM-HYP as well. I've just created https://github.com/seL4/l4v/issues/460 to help us remember for whenever there is some spare time.